### DISCUSSION TITLE
Slices should have the same capacity if the array is initialized with `cap`

### DISCUSSION THREAD

**Discussion by bstnbuck** [vlang/v#14635] [Ideas]
As shown in this bug report (#13572) it is not possible to pass the capacity of an array to a slice. However this seems not to be a bug in V. Go has this functionality, so that also an empty array can be specified with the outermost bound with the capacity. Just like this:

```go
// Go code
func main() {
	out := make([]byte, 0, 512)
	out = out[:cap(out)] // works

	fmt.Println(cap(out))
}
```
```v
// V code
fn main() {
	mut out := []byte{len: 0, cap: 512}
	out = out[..out.cap] // error <- out of bounds

	println(out.cap)
}
```
For example:
`copy` holds the capacity of the array (it copies a slice into that array). If using `out = out[0..].clone()` or simply `out = out[0..]` the capacity is `0` as below (it replaces the array with a slice of it).

```v
fn main() {
	mut out := []byte{cap: 512, init: 0}

	println(out.cap) //512
        copy(out, out[0..])
	println(out.cap) //512
	out = out[0..]
	println(out.cap) // <= 0
}
```

**Comment by ntrel** [comment]
It is unsafe to allow both the source array and the slice to append to the same memory because one of them could later be immutable after appending. If the other one is then appended to, the immutable one suddenly contains a different value.

However, it is possible to write a memory-safe function that takes the spare capacity from the original array, setting it back to `len`:
```v
// safe because `a` can no longer append
fn reuse_cap<T>(mut a []T) []T {
	r := unsafe {*a}
	a = a[..] // set `a.cap = a.len`
	return r
}

fn main() {
	mut a := [1,2,3]
	a << 4 // reallocate
	n := a.cap
	assert n > a.len
	b := reuse_cap(mut a)
	assert a.cap == a.len
	assert b.cap == n
}
```

**Comment by ntrel** [comment]
Just to mention, there is an existing function `trim` if you need to remove elements from the end of an array in-place, preserving `cap`:
https://modules.vlang.io/index.html#array.trim

I have made a pull to remove elements from the start of an array in-place, preserving `cap`: https://github.com/vlang/v/pull/13907

Using `arr.trim(end); arr.drop(start)` is equivalent to `arr[start..end]` but without changing `cap`.
